### ISSUE TITLE
Papyrus Compare is not specifically customized for AXCIOMA models

### ISSUE THREAD

**Issue by J17359** [ZeligsoftDev/CX4CBDDS#202]
# **Issue and tracking information**
### Developer's Actual time spent on fix (hours)  
# **Issue reporter to provide a detailed description of the issue in the space below**
Taken directly from https://github.com/ZeligsoftDev/CX4CBDDS/blob/papyrus/doc/compare/papyrus-compare-guide.md#limitations-and-known-issues:
> Papyrus Compare is not specifically customized for AXCIOMA models, which leads to showing several differences (and possibly conflicts) in the models that correspond to the underlying UML structure, rather than the AXCIOMA abstractions and constructs. While there are several filters that are applied to the differences views (and most are enabled by default), the tool does not yet include filters to show only the AXCIOMA structure.

We really need this to work for our users as it did back in RSA with ATCD. With our super simple BasicPubSub model merge example we've been using to work through configuration issues with @eposse in another ticket, we noticed that for the single module instantiation in the example, there are 10+ erroneous differences thrown in the merge tool. We only expect to see the one difference we actually made to the model. We do see that difference, but we also see the differences in the model that correspond to the underlying UML structure defined in the quote above. 

Our users' projects are **much** bigger than this BPS example and will contain a greater number of module instantiations in them. When they're doing their model merging, there's no way that we can direct them to ignore those sort of differences easily (i.e. no easy workaround). They'd have to know ahead of time what the model differences **should** be in order to know that they can ignore the differences shown that correspond to this underlying issue.

**Comment by eposse** [comment]
I think this is closely related to Issue #156.

**Comment by eposse** [comment]
@J17359 Should we give priority to this issue or to issue #170?

**Comment by J17359** [comment]
> @J17359 Should we give priority to this issue or to issue #170?

I'm of the opinion that both get done this year if possible (was going to ask for #170 after this one 😉). But I would give higher priority to **this** ticket (#202). We may be able to workaround the crashing by bumping the java heap space available in Papyrus so I'm more worried about this one that doesn't have a workaround at the moment.

**Comment by eposse** [comment]
Ok. I'll work on both issues in parallel, but will prioritize this one.

I'm linking Issue #193 here since we have a lot of info about this issue there.

**Comment by J17359** [comment]
@eposse - If you don't mind, please keep us in the loop every other day or so if possible as this is the most critical item blocking our release. 

I may be away on paternity leave any day now (wife is due Nov. 15th) but will do my best to sign in as often as I can to answer questions. @j26151 is also more than capable of answering questions/making decisions in my place.

**Comment by eposse** [comment]
@J17359 Sure. Right now I'm looking again at Issue #156 which, as I stated a few comments before, is strongly related to this issue. The heart of the issue seems to be related to instantiation of templates in a library, so I'm trying to understand how we are instantiating them and why they result in differences shown by Papyrus Compare.

**Comment by J17359** [comment]
> @J17359 Sure. Right now I'm looking again at Issue #156 which, as I stated a few comments before, is strongly related to this issue. The heart of the issue seems to be related to instantiation of templates in a library, so I'm trying to understand how we are instantiating them and why they result in differences shown by Papyrus Compare.

Thanks for the update @eposse !

This is totally up to @simonredding1 but we're all in to get @ysroh looking at this with you if you think that would help. SNA is hoping to release in early December to avoid conflicts with the holidays. We'd likely code freeze the week before (sometime around Thanksgiving here in the US). Having this figured out by then (if possible) would be awesome! Obviously, your tasking is approved by @simonredding1, I'm just providing SNA's "wish list" 😄

**Comment by simonredding1** [comment]
Hi Michael, 

 

Right now Young-Soo is occupied with some other work. If we do not make progress on the erroneous conflicts and out of memory problem soon, I will look at putting Young-Soo on this as well to get is resolved for you release. 

 

Regards, 

 

Simon

**Comment by J17359** [comment]
> Hi Michael, Right now Young-Soo is occupied with some other work. If we do not make progress on the erroneous conflicts and out of memory problem soon, I will look at putting Young-Soo on this as well to get is resolved for you release. Regards, Simon

We totally understand!

**Comment by eposse** [comment]
Hi @J17359 and @j26151,

Just a status report.

Young-Soo also looked at this issue and found the same problem I described in Issue #156. 

The summary is this: the models being compared have references to some elements in some library (in this case DDS_DCSP), and a textual comparison of the UML file shows the references to be identical in both left and right-hand sides, which is what makes it very strange. What happens is that when EMF Compare is trying to find the differences and runs into these references, it creates two dynamic instances of an EClass for the element in the library, and these DynamicEClass instances are actually proxies which have the URI to the library element. The two instances correspond to the left and right-hand sides. The problem is that EMF Compare is not smart enough to realize that the proxies are pointing to the exact same element, so it reports this as a difference.

So this looks like a bug in EMF Compare itself. 

One thing that we could try is to use a newer version of EMF Compare to see if it resolves the issue, but at this point I'm not certain about the consequences or possible side-effects. Furthermore, I think Papyrus Compare may have an upper bound on the EMF Compare version, but I'll verify that.

The alternative is to override the existing behaviour in EMF Compare via some extension point. So far I have not found an extension point to override the particular class that is responsible for this, but in the worse case, we can override the default "Difference Engine" of EMF Compare. It would be more work, but seems doable.

**Comment by J17359** [comment]
Thanks for the update @eposse! It certainly sounds like you've narrowed things down and have an idea on how to proceed. Looking forward to hearing from you again soon.

**Comment by eposse** [comment]
Hi @J17359 and @j26151 

Status update:

We have confirmed the diagnostics from the previous update and found also that EMF Compare is loading the library twice (actually three times for three-way comparisons) which explains the two instances of different EClasses and therefore the differences.

I've been trying to find the way to override this behaviour in EMF Compare. It seems possible, but it is quite convoluted, which is why it has taken me a couple of days to figure out, but once I find the final piece, the actual fix is really easy.

Other than that I've also worked on developing a proper target platform definition. This is something that doesn't affect you directly, but will ensure that the features and plugins in our development platform matches the built product, so that we can avoid the difficulties we discussed in issue #193.

**Comment by j26151** [comment]
@eposse This is great news, thank you for the hard work. Also, the idea of a proper target platform definition sounds like a really good idea.

**Comment by eposse** [comment]
Hi @j26151 and  @J17359

Just another update.

I finally found a way to override the problematic behaviour in EMF Compare and I'm implementing it.

I expect to have a build for you to test early next week.

**Comment by J17359** [comment]
That's great news @eposse! Thanks for being so persistent on this for us!

**Comment by eposse** [comment]
@J17359 and @j26151 

I've just pushed the fix to this issue (and #156).

I've tested it on CentOS and it works. For example the sample repository that you provided results in the following when merging branchB onto branchA:

![diff-merge](https://user-images.githubusercontent.com/3011155/99296213-642c1680-2814-11eb-88bd-be0e4165fcc2.png)

Note that the spurious differences are gone.

The resulting model after merging (adding all changes from right to left and saving) yields a valid model and it looks correct:

![diff-merge-result](https://user-images.githubusercontent.com/3011155/99296335-963d7880-2814-11eb-8b53-e311e355e295.png)

Can you try this build (branch name is papyrus-issue202)?

Thanks

**Comment by eposse** [comment]
Just an additional comment: while it resolves this issue, obviously other issues still may pop-up. In particular I noticed NPEs after merging a second time. These NPEs seem harmless (apparently just UI-related stuff), but they are a separate issue.

**Comment by j26151** [comment]
@eposse I have verified this fix. Thank you for figuring this out, this is huge for us. The NPE appear harmless on our end as well. We definitely want to investigate this more next year, but for now, this fix is great. Thank you.
